packages/components/src/custom-gradient-picker/control-points.js

```diff
@@ -0,0 +1,228 @@
+		// to another editor area.
+		event.stopPropagation();
+		const { gradientIndex, onChange, gradientAST } = this.props;
+		onChange( getGradientWithPositionAtIndexDecreased( gradientAST, gradientIndex ) );
+	}
+	render() {
+		const { children } = this.props;
+		return (
+			<KeyboardShortcuts shortcuts={ this.shortcuts }>
+				{ children }
+			</KeyboardShortcuts>
+		);
+	}
+}
+
+const ControlPointButton = withInstanceId(
+	function( {
+		instanceId,
+		isOpen,
+		position,
+		color,
+		onChange,
+		gradientIndex,
+		gradientAST,
+		...additionalProps
+	} ) {
+		const descriptionId = `components-custom-gradient-picker__control-point-button-description-${ instanceId }`;
+		return (
+			<ControlPointKeyboardMove
+				onChange={ onChange }
+				gradientIndex={ gradientIndex }
+				gradientAST={ gradientAST }
+			>
+				<Button
+					aria-label={
+						sprintf(
+							// translators: %1$s: gradient position e.g: 70%, %2$s: gradient color code e.g: rgb( 52,121,151).
+							__( 'Gradient control point at position %1$s with color code %2$s.' ),
+							position,
+							color
+						)
+					}
+					aria-describedby={ descriptionId }
+					aria-expanded={ isOpen }
+					className={
+						classnames(
+							'components-custom-gradient-picker__control-point-button',
+							{ 'is-active': isOpen }
+						)
+					}
+					style={ {
+						left: position,
+					} }
+					{ ...additionalProps }
+				/>
+				<div
+					className="screen-reader-text"
+					id={ descriptionId }>
+					{ __(
+						'Use your left or right arrow keys or drag&drop with the mouse to change the gradient position. Press the button to change the color or remove the control point.'
+					) }
+				</div>
+			</ControlPointKeyboardMove>
+		);
+	}
+);
+
+export default function ControlPoints( {
+	gradientPickerDomRef,
+	ignoreMarkerPosition,
+	markerPoints,
+	onChange,
+	gradientAST,
+	onStartControlPointChange,
+	onStopControlPointChange,
+} ) {
+	const controlPointMoveState = useRef();
+
+	const onMouseMove = ( event ) => {
+		const relativePosition = getHorizontalRelativeGradientPosition(
+			event.clientX,
+			gradientPickerDomRef.current,
+			GRADIENT_MARKERS_WIDTH,
+		);
+		const { gradientAST: referenceGradientAST, position, significantMoveHappened } = controlPointMoveState.current;
+		if ( ! significantMoveHappened ) {
+			const initialPosition = referenceGradientAST.colorStops[ position ].length.value;
+			if ( Math.abs( initialPosition - relativePosition ) >= MINIMUM_SIGNIFICANT_MOVE ) {
+				controlPointMoveState.current.significantMoveHappened = true;
+			}
+		}
+
+		if ( ! isControlPointOverlapping( referenceGradientAST, relativePosition, position ) ) {
+			onChange(
+				getGradientWithPositionAtIndexChanged( referenceGradientAST, position, relativePosition )
+			);
+		}
+	};
+
+	const onMouseUp = () => {
+		if ( window && window.removeEventListener ) {
+			window.removeEventListener( 'mousemove', onMouseMove );
+			window.removeEventListener( 'mouseup', onMouseUp );
```

Comment:
Should we also remove the listeners on unmount? Relying only on a pending `mouseup` event to take care of the cleanup feels fragile.